### PR TITLE
Transcript screen

### PR DESCRIPTION
**Pull request by bcaller**
Closes https://github.com/brave/brave-talk/issues/1409

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -53,18 +54,42 @@ export const App = () => {
     window.close();
   }
 
+  const initialiseRouteTranscript = () => {
```
Comment (DJAndries) [inline src/App.tsx:57]: nit: i know we're from the commonwealth, but US english is probably preferred :laughing:

File: src/App.tsx [hunk 2]
```diff
@@ -53,18 +54,42 @@ export const App = () => {
...
+    const match = window.location.pathname.match(
+      /^\/transcript-([a-z0-9]{50})$/,
+    );
+    if (match) {
+      return match[1];
+    }
+  };
+  const [routeTranscript, setRouteTranscript] = React.useState<
```
Comment (DJAndries) [inline src/App.tsx:65]: maybe `routeTranscriptId` instead?

**Review by DJAndries — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 3]
```diff
@@ -53,18 +54,42 @@ export const App = () => {
     window.close();
   }
 
+  const initialiseRouteTranscript = () => {
+    const match = window.location.pathname.match(
+      /^\/transcript-([a-z0-9]{50})$/,
```
[the remark this answers is not in the document]
Comment (bcaller) [reply src/App.tsx:59]: Something weird happens when there is a forward slash in the route: Visiting https://talk.brave.com/abc/def 302 redirects to https://talk.brave.com/

We would need to fix that first if we want forward slashes

File: src/App.tsx (hunk 1 above)
Comment (bcaller) [reply src/App.tsx]: Changed to `init`

File: src/App.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (DJAndries) [reply src/App.tsx:59]: ah ok. probably CloudFront related. perhaps it'd be better to address as a follow up

File: src/components/Transcript.tsx [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+      while (index > -1) {
+        const range = new Range();
+        range.setStart(textNode, index);
+        range.setEnd(textNode, index + str.length);
+        ranges.push(range);
+
+        index = content.indexOf(str, index + 1);
+      }
+    }
+    console.log(ranges);
```
Comment (DJAndries) [inline src/components/Transcript.tsx:112]: looks like this can be removed

File: src/components/Transcript.tsx [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+    const highlight = new Highlight(...ranges);
+    CSS.highlights.set("search-results", highlight);
+  }, [searchTerm]);
+
+  return (
+    <div
+      css={css`
+        margin: 93px auto;
```
Comment (DJAndries) [inline src/components/Transcript.tsx:120]: perhaps we should keep most styling in `styles` above, so the tree is easier to read

File: src/components/Transcript.tsx [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+              font-family: var(--Family-Headings, Poppins);
+              font-size: var(--Size-H2, 28px);
+              font-style: normal;
+              font-weight: 600;
+              line-height: var(--Line-height-H2, 36px);
+              letter-spacing: var(--Letter-spacing-Headings, -0.5px);
+              text-wrap: nowrap;
+            `}
+          >
+            Meeting Transcript
```
Comment (DJAndries) [inline src/components/Transcript.tsx:166]: we should use react-i18next for any copy. i.e. https://github.com/brave/brave-talk/blob/0621b711728796ddc921fe88205eab5237086909/src/components/JoinCallSection.tsx#L36-L38

File: src/components/Transcript.tsx [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+            justify-content: center;
+            align-items: center;
+            max-width: fit-content;
+          `}
+          kind="outline"
+          onClick={() => {
+            const link = document.createElement("a");
+            link.href = transcript.blobUrl;
+            link.download = `${transcript.id}.txt`;
+            link.click();
```
Comment (DJAndries) [inline src/components/Transcript.tsx:183]: any pitfalls with this approach? we could just add a `Content-Disposition` header in the transcript response

File: src/components/Transcript.tsx [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+              src={DownloadImage}
+              height="16"
+              width="18"
+              alt="download"
+              css={css`
+                margin-bottom: -2px;
+              `}
+            />
+          </span>
+          Download
```
Comment (DJAndries) [inline src/components/Transcript.tsx:197]: also i18n

File: src/components/Transcript.tsx [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+        range.setEnd(textNode, index + str.length);
+        ranges.push(range);
+
+        index = content.indexOf(str, index + 1);
+      }
+    }
+    console.log(ranges);
+    const highlight = new Highlight(...ranges);
+    CSS.highlights.set("search-results", highlight);
+  }, [searchTerm]);
```
Comment (DJAndries) [inline src/components/Transcript.tsx:115]: we may want to add `transcript` to the  list of dependencies

File: src/downloaded-transcript.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  for (const line of transcript.split("\n")) {
+    const timeOffset = line.match(/^(\d+m\d\ds) /)?.[1];
+    if (!timeOffset) continue;
+    const remainder = line.substring(timeOffset.length + 1);
+    for (const s in ACTION_NAMES) {
+      if (remainder.endsWith(s)) {
+        const result: TranscriptionEvent = {
+          timeOffset,
+          participant: remainder.substring(0, remainder.length - s.length),
+          message: s,
```
Comment (DJAndries) [inline src/downloaded-transcript.ts:47]: i18n as well

File: src/components/Transcript.tsx (hunk 6 above)
Comment (bcaller) [reply src/components/Transcript.tsx:166]: Do I need to add something in the JP translations, or do I just fill out the EN?

File: src/components/Transcript.tsx (hunk 7 above)
Comment (bcaller) [reply src/components/Transcript.tsx:183]: I am not aware of any issues with this. I believe this is the standard way to force a download of a file from javascript. The alternative is we need to make a change on the server-side with a parameter / route which adds the Content-Disposition header.

File: src/components/Transcript.tsx (hunk 6 above)
Comment (DJAndries) [reply src/components/Transcript.tsx:166]: I think English is fine for now

File: src/components/Transcript.tsx (hunk 7 above)
Comment (DJAndries) [reply src/components/Transcript.tsx:183]: okay. your call! will resolve for now since this seems okay

File: src/components/Transcript.tsx (hunk 6 above)
Comment (bcaller) [reply src/components/Transcript.tsx:166]: How do you trick it into displaying Japanese text anyway? I can't find a way

Comment (bcaller) [reply src/components/Transcript.tsx:166]: Oh the 日本語のlocalisation has been broken for I assume a while.

https://github.com/brave/brave-talk/blob/23497eb4c89a7194e3af499cf8b59d7bc9dd1a69/src/get-language-detector.ts#L5

when I'm pretty sure we need to also accept `=== "ja-JP"`... that is what devtools sets it to when you pretend to be in Tokyo.

My `navigator.language` is `en-GB` for instance and not simply `en`.

Comment (bcaller) [reply src/components/Transcript.tsx]: Fixed

image

**Review by DJAndries — APPROVED** [review]

**Pull request merged**